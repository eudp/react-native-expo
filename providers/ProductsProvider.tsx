import { createContext, useState, useEffect, ReactNode } from "react";
import { ActivityIndicator } from "react-native";
import { ProductType } from "../misc/types";
import { API_URL } from "../misc/constants";
import CustomText from "../components/CustomText";

const ProductsContext = createContext<ProductType[]>([]);

export const getProducts = async (): Promise<ProductType[]> => {
  const response = await fetch(API_URL);
  return response.json();
};

type Props = {
  children: ReactNode;
};

const ProductsProvider = ({ children }: Props) => {
  const [products, setProducts] = useState<ProductType[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  const fetchProducts = async () => {
    try {
      const data = await getProducts();
      setProducts(data);
      setIsLoading(false);
    } catch (error) {
      setError(error);
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchProducts();
  }, []);

  if (isLoading) {
    return <ActivityIndicator size="large" color="#000" />;
  }

  if (error) {
    return <CustomText>{error.message}</CustomText>;
  }

  return (
    <ProductsContext.Provider value={products}>
      {children}
    </ProductsContext.Provider>
  );
};

export { ProductsContext, ProductsProvider };
