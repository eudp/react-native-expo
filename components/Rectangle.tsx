import { useMemo, useContext } from "react";
import { StyleSheet, View } from "react-native";
import SectionTitle from "./SectionTitle";
import { ProductsContext } from "../providers/ProductsProvider";
import { getTotalPoints } from "../misc/utils";
import CustomText from "./CustomText";

const Rectangle = () => {
  const products = useContext(ProductsContext);

  const totalPoints = useMemo<number>(
    () => getTotalPoints(products),
    [products]
  );

  return (
    <View>
      <SectionTitle title="Tus Puntos" style={{ textTransform: "uppercase" }} />
      <View style={styles.containerRectangle}>
        <CustomText style={styles.topLeftText}>Diciembre</CustomText>
        <CustomText style={styles.centerText}>{totalPoints} pts</CustomText>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  containerRectangle: {
    backgroundColor: "#334FFA",
    alignItems: "center",
    justifyContent: "center",
    height: 143,
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.5,
    shadowRadius: 4,
    borderRadius: 20,
    marginLeft: 33,
    marginRight: 33,
    marginBottom: 20,
  },
  topLeftText: {
    position: "absolute",
    top: 21,
    left: 19,
    color: "#FFFFFF",
    fontSize: 16,
    fontFamily: "Avenir-Black",
    lineHeight: 22,
  },
  centerText: {
    color: "#FFFFFF",
    fontSize: 32,
    fontFamily: "Avenir-Black",
    lineHeight: 44,
  },
});

export default Rectangle;
