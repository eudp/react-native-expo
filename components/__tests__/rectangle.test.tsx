import React from "react";
import renderer from "react-test-renderer";
import { ProductsContext } from "../../providers/ProductsProvider";
import Rectangle from "../Rectangle";

describe("Rectangle", () => {
  it("renders correctly", () => {
    const products = [
      {
        image: "https://example.com/image.jpg",
        product: "Test Product",
        createdAt: "2022-04-09T12:34:56.000Z",
        is_redemption: false,
        points: 100,
        id: "1",
      },
      {
        image: "https://example.com/image.jpg",
        product: "Test Product",
        createdAt: "2022-04-09T12:34:56.000Z",
        is_redemption: true,
        points: 200,
        id: "3",
      },
    ];

    const tree = renderer
      .create(
        <ProductsContext.Provider value={products}>
          <Rectangle />
        </ProductsContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
