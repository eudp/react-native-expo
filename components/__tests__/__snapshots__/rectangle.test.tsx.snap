// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rectangle renders correctly 1`] = `
<View>
  <Text
    style={
      [
        {
          "fontFamily": "Avenir-Regular",
        },
        [
          {
            "color": "#9B9898",
            "fontFamily": "Avenir-Black",
            "fontSize": 14,
            "lineHeight": 19,
            "marginBottom": 15,
          },
          {
            "textTransform": "uppercase",
          },
        ],
      ]
    }
  >
    Tus Puntos
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#334FFA",
        "borderRadius": 20,
        "height": 143,
        "justifyContent": "center",
        "marginBottom": 20,
        "marginLeft": 33,
        "marginRight": 33,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.5,
        "shadowRadius": 4,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "color": "#FFFFFF",
            "fontFamily": "Avenir-Black",
            "fontSize": 16,
            "left": 19,
            "lineHeight": 22,
            "position": "absolute",
            "top": 21,
          },
        ]
      }
    >
      Diciembre
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "color": "#FFFFFF",
            "fontFamily": "Avenir-Black",
            "fontSize": 32,
            "lineHeight": 44,
          },
        ]
      }
    >
      -100
       pts
    </Text>
  </View>
</View>
`;
